### DISCUSSION TITLE
Vending Machine Support

### DISCUSSION THREAD

**Discussion by tosufever** [veryengineer/collab#17] [Ideas]
ok hear me out:

![914EBbjCLzL](https://user-images.githubusercontent.com/108495285/220220371-c3297153-9ca9-4937-81b8-9ef57ce6973b.jpg)

scope
- retrofit or refit an EXISTING bill exchanger.  start with the one that fits the most machines. 
- cram the node into the machine for just one, or
- use radioDoge to aggregate all machines in one building; or for (one) machines that are spread some distance (think hs football concession stand) and then
- at the other end of the radiodoge connection drop a pinodedoge.  (almost like a SCADA network for public works (reads your water & electric meter remotely)) honestly idk enough about lora radio works. 
- all transactions must be wallet to wallet on the l1. in the fastest & most user friendly method currently possible. no exceptions

just some quick thoughts:
- adding cc support to these vending machines has to be expensive. 
- they probably REQUIRE wired or dedicated wireless broadband access (and additional hardware +cellular service for remote machines) - i bet there are some vending machine shibes out there we could get to chime in? anyone?
- this would maybe give machine owners a "free" (by installing the hw) digital alternative to credit cards. 
- it eliminates several reoccurring services, and probably a big % savings in tx cost & fees.  
- if you can mesh it... "scada as a service :-)" there are several "small business' in a box" opportunities here (that seem scalable easily from 0) 
- a node in a vending machine would be sticky. 

discuss.

inspired by maximus' ideas thread & a twitter thread.
https://github.com/veryengineer/collab/discussions/12#discussioncomment-5056290
https://twitter.com/K1ngSatoshee/status/1627826253091971073
radiodoge twitter - @michilumin
*can't link radiodoge, because the @michilumin thread was about temporarily removing it from github.

**Comment by patricklodder** [comment]
Do owners of vending machines often have one or many of these?

**Reply by K1ngsatoshee** [reply]
Most own a plethora. Thanks for adding this!

**Reply by tosufever** [reply]
many small business own their own machines. next to the checkout counter, in the lobby, in the break room. one, or a handful. stocked from sams club.

**Reply by patricklodder** [reply]
>Most own a plethora

In that case, running a node inside the vending machine isn't really cost efficient. All you need is a local bip32 extended public key, a counter and a display to put a bip21 QR code. Then all you need is a socket to, say, an electrumx server and you simply poll the address for incoming moneys. and done.

This means you can probably have a 1MB runtime requirement or so on the device. And if you're just piloting it, you can use my public electrum server for comms so you don't have to run a 12GB RAM server and save yourself a hundred bucks a month.

**Reply by patricklodder** [reply]
> many small business own their own machines. next to the checkout counter, in the lobby, in the break room. one, or a handful. stocked from sams club.

Own or lease?

**Reply by tosufever** [reply]
at the smallest scale, one or two machines, it's hard or impossible to even get a machine leased, delivered & serviced.  so the one in the lobby of a non-chain retail / service place (local hardware store) is probably (2nd hand) owned & stocked by a trip to sams club.

**Reply by patricklodder** [reply]
Okay. So if this is the only use case for accepting Dogecoin for an entity then, yes. I'd still put an SPV node in there that can sync from a checkpoint rather than a full node, or even a pruning node, for the simple reason that you don't want hours of downtime for a vending machine while it syncs. So with the tools we have, what you'd need is something that is a little like a phone: display + 1GB RAM + something that runs java efficiently (i.e. android)

You won't need a lot of disk space because for SPV you simply want to store the last... n-thousand headers and prune as you go, only keep blocks that contain the tx you need.

**Comment by michilumin** [comment]
Sorry for slow response, my GH notifications still only come through email. 9_9 .. Anyways --

I do think that RD could/would be applicable here. 

Some of the pi-like Rockchip 3xxx based ARM SBCs that I'm currently fitting for Radidoge could absolutely handle this amount of data and processing; RD could relay around a "campus" pretty transparently; RD could act as sort of command and control for a SPV node even; if chatter in an area was to be avoided.  And with message relay then at least ubiquitous wifi coverage over said "campus" would not be needed.

If we have a SPV node ready to go out there I'd love to try some remote transactions and see if we can do proof of concept. I know something was in the works but I haven't been keeping up with it.

**Reply by tosufever** [reply]
i'm still getting familiar with github - it looks like i replied in the wrong box. 
https://github.com/veryengineer/collab/discussions/17?sort=new#discussioncomment-5095163

**Reply by patricklodder** [reply]
Happens me all the time.

**Comment by tosufever** [comment]
too afraid to ask, i was trying to figure out what an spv node even did the other day. i'm still not entirely sure. bookmarked the following:
https://github.com/ShibeTechnology/dogecoin-spv-node
https://github.com/BitcoinAmiens/dogecoin-spv-node
https://github.com/dogecoin/libdohj 


for anyone else who is curious there is a good description here: https://www.bitdegree.org/crypto/learn/crypto-terms/what-is-simplified-payment-verification-spv

**Reply by patricklodder** [reply]
The first 2 are the same thing, the one under ShibeTechnology is the most recent one and has some payment channel features. 

Libdohj is a library that would need an implementation (not hard, but java), but that library is now unmaintained and in the future, libdogecoin should be a good replacement. I know that @edtubbs and @xanimo were working on something there.

I was discussing with @jimjimmiejames an hour ago or so and promised that I'll work on some simple prototypes using electrum that will allow for things like listening on a web socket to receive updates about an address. Am going to do the first iteration in js so that it also works in a browser but nodejs is not lean - so that's nice for a PoC but not for a final productized thing. Can re-wire btcd/wire too.. that isn't hard.
